### ISSUE TITLE
🐛 BUG: Wrangler chooses today's date as compatibility date instead of latest available, if none is specified by user

### ISSUE THREAD

**Issue by mxschumacher** [cloudflare/workers-sdk#3639]
### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.3.0

### What version of Node are you using?

18.16.1

### What operating system are you using?

Linux (Fedora 38)

### Describe the Bug

TLDR: If the compability_date is **not specified** in wrangler.toml or passed as a command line argument, the "fallback date" should be the latest available compatibility_date and not the latest date (e.g. today). As it currently stands, the CLI basically disagrees with itself (please see full output below). 

My wrangler.toml file does not contain an entry for compatibility_date

At the time of writing, the date is "2023-07-19"

As of today: 
> "The latest compatibility date supported by the installed Cloudflare Workers Runtime is "2023-07-17"

In a newly started project (created with `npx wrangler init .`), I ran the following command:

> npx wrangler dev src/index.js

Please see the full output here:

```
npx wrangler dev src/index.js                                                                 
 ⛅️ wrangler 3.3.0
------------------
wrangler dev now uses local mode by default, powered by 🔥 Miniflare and 👷 workerd.
To run an edge preview session for your Worker, use wrangler dev --remote
▲ [WARNING] No compatibility_date was specified. Using today's date: 2023-07-19.

  Add one to your wrangler.toml file:
  
  compatibility_date = "2023-07-19"
  
  or pass it in your terminal:
  
  --compatibility-date=2023-07-19
  
  See https://developers.cloudflare.com/workers/platform/compatibility-dates/ for more information.

⎔ Starting local server...
[mf:wrn] The latest compatibility date supported by the installed Cloudflare Workers Runtime is "2023-07-17",
but you've requested "2023-07-19". Falling back to "2023-07-17"...
[mf:inf] Ready on http://127.0.0.1:8787/
```

**Comment by Cherry** [comment]
I'm not sure I'd personally consider this one a bug, but would instead consider the warning as just bad DX.

Wisdom and advice for _years_ has been to use today's date as your compat date when starting a new project, going back to the original inception of Workers and still on the docs today. The introduction of "supported" dates and this warning is the weird part to me.

I opened an issue at https://github.com/cloudflare/workers-sdk/issues/3363 discussing this more.

**Comment by petebacondarwin** [comment]
This was resolved in https://github.com/cloudflare/workers-sdk/pull/3343, which was released into wrangler@3.1.2 and create-cloudflare@2.0.11, for new projects created by `npx wrangler init`.

I just confirmed that this is the case:

```
npx wrangler@latest init . -y
 ⛅️ wrangler 3.3.0
------------------
Using npm as package manager.
▲ [WARNING] The `init` command is no longer supported. Please use `npm create cloudflare@2 -- . --wrangler-defaults` instead.

  The `init` command will be removed in a future version.


Running `npm create cloudflare@2 -- . --wrangler-defaults`...

using create-cloudflare version 2.0.13

╭ Create an application with Cloudflare Step 1 of 3
│ 
├ In which directory do you want to create your application?
│ dir ./.
│
├ What type of application do you want to create?
│ type "Hello World" Worker
│
├ Copying files from "hello-world" template
│ 
├ Retrieving current workerd compatibility date 
│ compatibility date 2023-07-17
│ 
├ Do you want to use git for version control?
│ yes git
│
╰ Application created 

╭ Installing dependencies Step 2 of 3
│ 
├ Installing dependencies 
│ installed via `npm install`
│ 
├ Committing new files 
│ git initial commit
│ 
╰ Dependencies Installed 

╭ Deploy with Cloudflare Step 3 of 3
│ 
├ Do you want to deploy your application?
│ no deploy via `npm run deploy`
│
├  APPLICATION CREATED  Deploy your application with npm run deploy
│ 
│ Navigate to the new directory: cd foooo
│ Run the development server npm run start
│ Deploy your application npm run deploy
│ Read the documentation https://developers.cloudflare.com/workers
│ Stuck? Join us at https://discord.gg/cloudflaredev
│ 
╰ See you again soon! 
```

@mxschumacher can you output the logs from your use of `npx wrangler init`? Perhaps it didn't use the latest versions of Wrangler and C3?

**Comment by mxschumacher** [comment]
Hello @petebacondarwin , please note that my bug report is about what happens after I run

> npx wrangler dev src/index.js

Meaning any project, not necessarily generated by C3 or similar, that does not have the compatibility_date defined in the wrangler .toml

I am using wrangler 3.3.0, you can see it at the very top of the stack trace I posted.

**Comment by petebacondarwin** [comment]
Oh! Sorry I missed this line in the output:

```
▲ [WARNING] No compatibility_date was specified. Using today's date: 2023-07-19.
```

and that you had removed the compatibility date from the wrangler.toml since I read these lines:

> In a newly started project (created with npx wrangler init .), I ran the following command:
> ```
> npx wrangler dev src/index.js
> ```

And assumed that you were using the output of `npx wrangler init` directly.

Updated the title of the issue to be more explicit about this.

**Comment by petebacondarwin** [comment]
I think that the correct resolution is what I suggest in https://github.com/cloudflare/workers-sdk/issues/3363#issuecomment-1566106557. I.e. we only warn if the local version of workerd is not latest and the date is too recent. Then for the vast majority projects this will not show a warning, and when it does it is a useful warning.

**Comment by admah** [comment]
The fix for this should be available in Wrangler 3.15.0 (#4209). Could you retest and let us know if the issue is resolved?

**Comment by penalosa** [comment]
This was fixed in #4209

**Comment by ffxsam** [comment]
> This was fixed in [#4209](https://github.com/cloudflare/workers-sdk/pull/4209)

I don't think so. Just got it today.

```
19:44:31 [ERROR] [vite] service core:user:reelcrafter-marketing: This Worker requires compatibility date "2026-08-13", but the newest date supported by this server binary is "2026-08-11".
The Workers runtime failed to start. There was likely a problem with the workerd binary or your configuration.
Runtime stderr:
{"timestamp":"1786581871666","level":"error","source":"src/workerd/server/json-logger.c++:127","message":"service core:user:reelcrafter-marketing: This Worker requires compatibility date \"2026-08-13\", but the newest date supported by this server binary is \"2026-08-11\".","context_depth":0}
  Stack trace:
    at StartupLogBuffer.handleStartupFailure (/Users/samh/Sites/reelcrafter/reelcrafter-marketing/node_modules/.pnpm/miniflare@5.20260804.1-alpha/node_modules/miniflare/dist/src/index.js:97695:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:104:5)
    at async Mutex.runWith (/Users/samh/Sites/reelcrafter/reelcrafter-marketing/node_modules/.pnpm/miniflare@5.20260804.1-alpha/node_modules/miniflare/dist/src/index.js:85641:48)
    at async _Miniflare.dispatchFetch (/Users/samh/Sites/reelcrafter/reelcrafter-marketing/node_modules/.pnpm/miniflare@5.20260804.1-alpha/node_modules/miniflare/dist/src/index.js:112279:5)
    at async Promise.all (index 0)
```

**Comment by workers-devprod** [comment]
> [!NOTE]
> This is an automated, advisory triage report generated by workers-devprod. It is not an official maintainer response — a maintainer will follow up.

<details>
<summary>🤖 Automated triage report</summary>

# Issue Triage: cloudflare/workers-sdk#3639

## Summary

When `compatibility_date` is not specified, Wrangler falls back to today's date, which can be newer than the latest compatibility date supported by the installed runtime — producing a self-contradictory warning.

## Classification

- **Type:** Bug
- **Component:** `packages/wrangler`
- **Severity:** Low
- **Has Reproduction:** Yes (clear steps + full console output)
- **Quality:** Complete

## Findings

- **Created:** not readable from `context.json` (see note below)
- **Author:** mxschumacher (Max Schumacher)
- **Version:** wrangler 3.3.0, Node 18.16.1, Linux (Fedora 38)
- **Labels:** not readable from `context.json` (see note below)
- **Comments:** not readable from `context.json` (see note below)

> **Data note:** `context.json` is a single-line JSON document whose body field exceeds the 2000-character read/grep truncation limit. The issue body was fully readable, but trailing metadata (labels, comment count, created/updated dates, and state) fall past the truncation boundary and could not be extracted with the available read/grep/glob tools. The analysis below is based on the fully-readable issue body.

### Analysis

- **Component identification:** This is a Wrangler tooling issue, not a runtime issue. The fallback-date selection logic (defaulting a missing `compatibility_date` to today's date and emitting the "Using today's date" warning) lives in Wrangler, so it belongs in `packages/wrangler`.
- **The bug is legitimate and well-described:** Wrangler defaults an unspecified `compatibility_date` to today's date, then Miniflare/workerd warns that the requested date exceeds the latest date the installed runtime supports (2023-07-19 requested vs. 2023-07-17 supported). The two warnings contradict each other, which is confusing for new users following `wrangler init`.
- **Proposed fix is reasonable:** Falling back to the latest compatibility date the installed runtime actually supports (rather than `Date.now()`) would eliminate the contradiction. This is a small, contained change but has some subtlety (local vs. remote dev may want different fallbacks, and the fallback must be discoverable from the bundled runtime).
- **Reproduction quality:** Complete — includes exact command, full annotated output, versions, and OS. No missing-info gap.
- **Age consideration:** Reported against wrangler 3.3.0 (mid-2023). The default-compatibility-date warning behavior and messaging have likely changed across the many wrangler releases since. This warrants verification against the current version before committing to a fix, as the behavior may already have been adjusted.

## Recommendation

**Status:** NEEDS VERIFICATION

**Reasoning:** The bug is valid and clearly reproducible as reported, but it was filed against wrangler 3.3.0 which is many releases old; the default-date fallback logic and warning messaging may have changed since. A maintainer should confirm whether the self-contradictory warning still occurs on the current version before scheduling a fix.

**Action:** Reproduce on the latest wrangler with a config that omits `compatibility_date`. If the contradictory warning persists, fix the fallback in `packages/wrangler` to clamp/default to the latest compatibility date supported by the bundled runtime instead of today's date. If already resolved, close as fixed.

**Suggested Labels:** _(omitted — repository label set could not be confirmed from the truncated `context.json`)_

### Suggested Comment

> Thanks for the detailed report — the contradiction between "Using today's date" and the runtime's "latest supported compatibility date" warning is a real UX problem. This was filed against wrangler 3.3.0, and the default-compatibility-date handling has changed a fair amount since then. Could you confirm whether you still see this contradictory warning on the latest version of wrangler? If it's still happening, we'll look into defaulting the fallback to the latest compatibility date the installed runtime supports rather than today's date.

</details>

<details>
<summary>Structured summary (JSON)</summary>

```json
{
  "issueNumber": 3639,
  "title": "compatibility_date fallback should use the latest available date, not today's date",
  "githubUrl": "https://github.com/cloudflare/workers-sdk/issues/3639",
  "recommendation": "NEEDS VERIFICATION",
  "difficulty": "medium",
  "reasoning": "Valid, clearly reproducible bug where an unspecified compatibility_date defaults to today's date, contradicting the runtime's latest-supported-date warning. Filed against wrangler 3.3.0, so it needs verification on the current version since the fallback logic may have changed.",
  "suggestedAction": "Reproduce on the latest wrangler with a config that omits compatibility_date; if the contradictory warning persists, fix the fallback in packages/wrangler to use the latest compatibility date supported by the bundled runtime instead of today's date, otherwise close as fixed.",
  "hasSuggestedComment": true,
  "suggestedLabels": []
}

```

</details>
